From 36a855fed6e25c9ea037050188311d8fc70dc98b Mon Sep 17 00:00:00 2001
From: David Sniatkiwicz <davidsn@marvell.com>
Date: Sun, 21 Aug 2016 09:13:13 +0300
Subject: [PATCH 0440/1345] fix: thermal: ap806: fix thermal overheat
 threshold to celsius calculation

commit  517974393ff3056cebbd419565b1a87d46753f6d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Due to a wrong overheat threshold print values,
this patch fixes threshold to temperature(celsius) conversion,
by adding a condition that checks the threshold value,
and reverts the 2s complement, which calculated by
ap806_thresh_val_calc() function(when the value
bigger than TSEN max value(511)).

Change-Id: I6701c2bd06a41b710e21ce5054b4d4ef0cfdc3d4
Signed-off-by: David Sniatkiwicz <davidsn@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/32089
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Haim Boot <hayim@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/thermal/armada_thermal.c |   12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/thermal/armada_thermal.c b/drivers/thermal/armada_thermal.c
index ac9c162..bdd2445 100644
--- a/drivers/thermal/armada_thermal.c
+++ b/drivers/thermal/armada_thermal.c
@@ -151,11 +151,19 @@ inline unsigned int ap806_thresh_val_calc(unsigned int celsius_temp,
 	return thresh_val & data->temp_mask;
 }
 
-inline unsigned int ap806_thresh_celsius_calc(int thresh_val,
-					 int hyst, struct armada_thermal_data *data)
+inline unsigned int ap806_thresh_celsius_calc(int thresh_val, int hyst,
+					      struct armada_thermal_data *data)
 {
 	unsigned int mcelsius_temp;
 
+	/* This calculation is the opposite to the calculation done in
+	 * ap806_thresh_val_calc(). In order to get the right result,
+	 * need to revert the 2s complement calculation,
+	 * which occurs when threshold bigger then TSEN max value(511).
+	 */
+	if (thresh_val >= AP806_TSEN_OUTPUT_MSB)
+		thresh_val -= AP806_TSEN_OUTPUT_COMP;
+
 	mcelsius_temp = (((data->coef_m * (thresh_val + hyst)) +
 			  data->coef_b) / data->coef_div);
 
-- 
1.7.9.5

